File: bench/diff_memweight.c
// SPDX-License-Identifier: GPL-2.0
// Tier-2.5 differential oracle: C original vs Rust translation, memweight.
// Reference extracted from lib/memweight.c (v7.1); kept byte-identical.
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned long ul;
#define BITS_PER_LONG (sizeof(ul) * 8)
#define INT_MAX 2147483647

static unsigned int hweight8(unsigned char w)
{
	unsigned int c = 0;
	while (w) { c += w & 1; w >>= 1; }
	return c;
}

static unsigned int bitmap_weight(const ul *bitmap, unsigned int nbits)
{
	unsigned int n = nbits / BITS_PER_LONG;
	unsigned int rem = nbits % BITS_PER_LONG;
	unsigned int c = 0;
	for (unsigned int i = 0; i < n; i++) {
		ul w = bitmap[i];
		while (w) { c += w & 1; w >>= 1; }
	}
	if (rem) {
		ul w = bitmap[n] & ((rem == BITS_PER_LONG) ? ~0UL : ((1UL << rem) - 1));
		while (w) { c += w & 1; w >>= 1; }
	}
	return c;
}

static size_t memweight(const void *ptr, size_t bytes)
{
	size_t ret = 0;
	size_t longs;
	const unsigned char *bitmap = ptr;

	for (; bytes > 0 && ((unsigned long)bitmap) % sizeof(long);
			bytes--, bitmap++)
		ret += hweight8(*bitmap);

	longs = bytes / sizeof(long);
	if (longs) {
		if (longs >= (size_t)(INT_MAX / BITS_PER_LONG)) {
			fprintf(stderr, "BUG_ON hit (test bug, size too large)\n");
			abort();
		}
		ret += bitmap_weight((unsigned long *)bitmap,
				longs * BITS_PER_LONG);
		bytes -= longs * sizeof(long);
		bitmap += longs * sizeof(long);
	}
	for (; bytes > 0; bytes--, bitmap++)
		ret += hweight8(*bitmap);

	return ret;
}

// Explicit LCG (same constants as bench/diff_base64.c).
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

#define MAXLEN 200

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 271828;

	// Over-allocate and offset the start within a larger buffer so
	// `bitmap`'s initial alignment (relative to sizeof(long)) varies
	// across iterations, same as real callers with arbitrary pointers.
	unsigned char storage[MAXLEN + 16];

	for (long i = 0; i < n; i++) {
		int offset = lcg_next() % 8; // vary alignment 0..7
		int len = lcg_next() % (MAXLEN - offset);
		for (int k = 0; k < len; k++) storage[offset + k] = (unsigned char)lcg_next();

		size_t w = memweight(storage + offset, (size_t)len);
		printf("weight,%d,%d,%zu\n", offset, len, w);
	}
	return 0;
}

File: bench/diff_memweight.rs
// SPDX-License-Identifier: GPL-2.0
//! Tier-2.5 differential oracle: Rust translation side, memweight.
//! Faithful copy of lib/memweight_rs.rs's algorithm (kernel-crate
//! bindings/export stripped) — same protocol/LCG as diff_memweight.c.

const BITS_PER_LONG: usize = usize::BITS as usize;
const SIZEOF_LONG: usize = core::mem::size_of::<usize>();
const INT_MAX: usize = i32::MAX as usize;

fn hweight8(mut w: u8) -> u32 {
    let mut c = 0u32;
    while w != 0 {
        c += (w & 1) as u32;
        w >>= 1;
    }
    c
}

fn bitmap_weight(bitmap: &[usize], nbits: u32) -> u32 {
    let n = nbits as usize / BITS_PER_LONG;
    let rem = nbits as usize % BITS_PER_LONG;
    let mut c = 0u32;
    for &w in &bitmap[..n] {
        c += w.count_ones();
    }
    if rem != 0 {
        let mask = if rem == BITS_PER_LONG {
            usize::MAX
        } else {
            (1usize << rem) - 1
        };
        c += (bitmap[n] & mask).count_ones();
    }
    c
}

fn memweight(ptr: *const u8, mut bytes: usize) -> usize {
    let mut ret: usize = 0;
    let mut bitmap = ptr;

    while bytes > 0 && (bitmap as usize) % SIZEOF_LONG != 0 {
        let b = unsafe { *bitmap };
        ret += hweight8(b) as usize;
        bytes -= 1;
        bitmap = unsafe { bitmap.add(1) };
    }

    let longs = bytes / SIZEOF_LONG;
    if longs != 0 {
        if longs >= INT_MAX / BITS_PER_LONG {
            panic!("BUG_ON hit (test bug, size too large)");
        }
        // SAFETY: `bitmap` is SIZEOF_LONG-aligned here (host allocation is
        // byte-granular so this cast needs read_unaligned in general, but
        // the loop above guarantees alignment before this point).
        let words = unsafe {
            core::slice::from_raw_parts(bitmap as *const usize, longs)
        };
        ret += bitmap_weight(words, (longs * BITS_PER_LONG) as u32) as usize;
        bytes -= longs * SIZEOF_LONG;
        bitmap = unsafe { bitmap.add(longs * SIZEOF_LONG) };
    }

    while bytes > 0 {
        let b = unsafe { *bitmap };
        ret += hweight8(b) as usize;
        bytes -= 1;
        bitmap = unsafe { bitmap.add(1) };
    }

    ret
}

// Identical LCG to diff_memweight.c.
struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

const MAXLEN: usize = 200;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(271828);
    let mut rng = Lcg(seed);

    let mut storage = [0u8; MAXLEN + 16];

    for _ in 0..n {
        let offset = (rng.next() % 8) as usize;
        let len = (rng.next() as usize) % (MAXLEN - offset);
        for k in 0..len {
            storage[offset + k] = (rng.next() & 0xff) as u8;
        }

        let w = memweight(unsafe { storage.as_ptr().add(offset) }, len);
        println!("weight,{},{},{}", offset, len, w);
    }
}
